debugging.md

```diff
@@ -0,0 +1,107 @@
+# Debugging
+
+This document aims to collect info on the current status and future opportunities for debugging Rust.
+
+## Debugging architecture overview
+
+The compiler generates debuginfo for Rust programs (see https://github.com/rust-lang/rust/tree/master/src/librustc_trans/debuginfo).
+This is an LLVM-specific thing. LLVM then processes this through optimisations
+(with varying degrees of success) and outputs debug symbols with the library.
+These are DWARF for the Linux, Mac, and Windows/GNU toolchains, and pdb (program
+database) for Windows/visual studio (pdb support is WIP in LLVM, I believe).
+
+Debuggers consume the generated symbols. Visual Studio only works on Windows.
+LLDB works everywhere but is primarily used as a backend for XCode on MacOS. GDB
+works everywhere. GDB and LLDB are command
+line tools, but can have frontends.
+
+Debugger support can come in various forms in this architecture:
+* we could invent our own debugger and debug symbols (this is infeasible due to resourcing)
+* generate better debuginfo in the compiler - most of the low hanging fruit here is done
+* extend debug symbols with Rust-specific forms
```

Comment:
DWARF is designed to support vendor-specific attributes on individual debugging entries, and even vendor-specific debugging entry types. DWARF ensures that consumers (i.e. debuggers) can skip over information that they don't understand. So the format itself is definitely ready to be extended in a clean way.

You'd have to get LLVM to emit the information; I have no idea what's entailed there.

You'd have to get GDB and LLDB to use the information. We've been able to get changes into GDB in the past, so that's promising. LLDB, I don't know anything about.